File: topics/mpi/practicals/02.MPI_pt2pt/fortran/5.exchange_ghost_cells_solution.f90
PROGRAM ghost_cell_exchange

! process decomposition on 4*4 grid, ranks are connected in a cyclic manner
! for instance, rank 0 and 12 are connected
!  |-----------|
!  | 0| 4| 8|12|
!  |-----------|
!  | 1| 5| 9|13|
!  |-----------|
!  | 2| 6|10|14|
!  |-----------|
!  | 3| 7|11|15|
!  |-----------|

! Each process works on a 10*10 (SUBDOMAIN) block of data
! the d corresponds to data, g corresponds to "ghost cells"
! and x are empty (not exchanged for now)

!  xggggggggggx
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  gddddddddddg
!  xggggggggggx

! task: each rank has to find its left and right neighbor and
! send them the data they need (left array goes to left neighbor
!                               and right array goes to right neighbor)

  USE MPI
  IMPLICIT NONE

  INTEGER SUBDOMAIN, DOMAINSIZE
  PARAMETER (SUBDOMAIN = 10)
  PARAMETER (DOMAINSIZE = SUBDOMAIN+2)
  INTEGER rank, size, i, j, rank_right, rank_left, ierror
  DOUBLE PRECISION data(DOMAINSIZE,DOMAINSIZE)
  INTEGER request
  INTEGER status(MPI_STATUS_SIZE)

  CALL MPI_Init(ierror)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierror)
  CALL MPI_Comm_size(MPI_COMM_WORLD, size, ierror)

  IF (size.NE.16) THEN
     WRITE (*,*)"please run this with 16 processors"
     CALL MPI_Finalize(ierror);
     STOP
  END IF
  DO i=1, DOMAINSIZE, 1
    DO j=1, DOMAINSIZE, 1
      data(i,j)=rank
    END DO
  END DO

  rank_right=mod(rank+4,16)
  rank_left=mod(rank+16-4,16)

  !  ghost cell exchange with the neighbouring cells (cyclic) on the left and on the right using:
  !  a) MPI_Send, MPI_Irecv
  !  b) MPI_Isend, MPI_Recv
  !  c) MPI_Sendrecv

  !  to the left

  ! a)
  CALL MPI_Irecv(data(2,DOMAINSIZE), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_right, 0, MPI_COMM_WORLD, request, ierror)
  CALL MPI_Send(data(2,2), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_left, 0, MPI_COMM_WORLD, ierror)
  CALL MPI_Wait(request, status, ierror)
  ! b)
  CALL MPI_Isend(data(2,2), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_left, 0, MPI_COMM_WORLD, request, ierror)
  CALL MPI_Recv(data(2,DOMAINSIZE), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_right, 0, MPI_COMM_WORLD, status, ierror)
  CALL MPI_Wait(request, status, ierror)
  ! c)
  CALL MPI_Sendrecv(data(2,2), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_left, 0, data(2,DOMAINSIZE), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_right, 0, MPI_COMM_WORLD, status, ierror)
  !  to the right
  ! a)
  CALL MPI_Irecv(data(2,1), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_left, 0, MPI_COMM_WORLD, request, ierror)
  CALL MPI_Send(data(2,DOMAINSIZE-1), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_right, 0, MPI_COMM_WORLD, ierror)
  CALL MPI_Wait(request, status, ierror);
  ! b)
  CALL MPI_Isend(data(2,DOMAINSIZE-1), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_right, 0, MPI_COMM_WORLD, request, ierror)
  CALL MPI_Recv(data(2,1), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_left, 0, MPI_COMM_WORLD, status, ierror)
  CALL MPI_Wait(request, status, ierror);
  ! c)
  CALL MPI_Sendrecv(data(2,DOMAINSIZE-1), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_right, 0, data(2,1), SUBDOMAIN, MPI_DOUBLE_PRECISION, rank_left, 0, MPI_COMM_WORLD, status, ierror)

  IF (rank.EQ.9) THEN
     WRITE (*,*) 'data of rank 9 after communication'
     DO i=1, DOMAINSIZE, 1
        DO j=1, DOMAINSIZE, 1
          WRITE (*,'(F6.1)',advance='no') data(i,j)
        END DO
        WRITE (*,*)
     END DO
  END IF

  CALL MPI_Finalize(ierror);

END PROGRAM

File: topics/mpi/solutions/02.MPI_pt2pt/cxx/5.exchange_ghost_cells_solution.c
/****************************************************************
 *                                                              *
 * This file has been written as a sample solution to an        *
 * exercise in a course given at the CSCS Summer School.        *
 * It is made freely available with the understanding that      *
 * every copy of this file must include this header and that    *
 * CSCS take no responsibility for the use of the enclosed      *
 * teaching material.                                           *
 *                                                              *
 * Purpose: Exchange ghost cell in 2 directions                 *
 *                                                              *
 * Contents: C-Source                                           *
 *                                                              *
 ****************************************************************/

/* Use only 16 processes for this exercise
 * Send the ghost cell in two directions: top to bottom and bottom to top
 * consider cyclic boundaries, for instance, rank 0 and rank 12 are connected together
 *
 * process decomposition on 4*4 grid
 *
 *  |-----------|
 *  | 0| 1| 2| 3|
 *  |-----------|
 *  | 4| 5| 6| 7|
 *  |-----------|
 *  | 8| 9|10|11|
 *  |-----------|
 *  |12|13|14|15|
 *  |-----------|
 *
 * Each process works on a 10*10 (SUBDOMAIN) block of data
 * the D corresponds to data, g corresponds to "ghost cells"
 * xggggggggggx
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * gDDDDDDDDDDg
 * xggggggggggx
 */

/* Tasks:
 * A. each rank has to find its top and bottom neighbor
 * B. send them the data they need
 *    - top array goes to top neighbor
 *    - bottom array goes to bottom neighbor
 */

#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

#define SUBDOMAIN 10
#define DOMAINSIZE (SUBDOMAIN+2)

int main(int argc, char *argv[])
{
    int rank, size, i, j, rank_bottom, rank_top;
    double data[DOMAINSIZE*DOMAINSIZE];
    MPI_Request request;
    MPI_Status status;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    if (size!=16) {
        printf("please run this with 16 processors\n");
        MPI_Finalize();
        exit(1);
    }

    for (i=0; i<DOMAINSIZE*DOMAINSIZE; i++) {
        data[i]=rank;
    }

    rank_bottom=(rank+4)%16;
    rank_top=(rank+16-4)%16;


    //  ghost cell exchange with the neighbouring cells (cyclic) to the bottom and to the top using:
    //  a) MPI_Send, MPI_Irecv, MPI_Wait
    //  b) MPI_Isend, MPI_Recv, MPI_Wait
    //  c) MPI_Sendrecv

    //  to the top

    // a)
    MPI_Irecv(&data[2-1+(DOMAINSIZE-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_bottom, 0, MPI_COMM_WORLD, &request);
    MPI_Send(&data[2-1+(2-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_top, 0, MPI_COMM_WORLD);
    MPI_Wait(&request, &status);

    // b)
    MPI_Isend(&data[2-1+(2-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_top, 0, MPI_COMM_WORLD, &request);
    MPI_Recv(&data[2-1+(DOMAINSIZE-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_bottom, 0, MPI_COMM_WORLD, &status);
    MPI_Wait(&request, &status);

    // c)
    MPI_Sendrecv(&data[2-1+(2-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_top, 0, &data[2-1+(DOMAINSIZE-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_bottom, 0, MPI_COMM_WORLD, &status);

    //  to the bottom
    // a)
    MPI_Irecv(&data[2-1+(1-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_top, 0, MPI_COMM_WORLD, &request);
    MPI_Send(&data[2-1+(DOMAINSIZE-1-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_bottom, 0, MPI_COMM_WORLD);
    MPI_Wait(&request, &status);

    // b)
    MPI_Isend(&data[2-1+(DOMAINSIZE-1-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_bottom, 0, MPI_COMM_WORLD, &request);
    MPI_Recv(&data[2-1+(1-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_top, 0, MPI_COMM_WORLD, &status)    ;
    MPI_Wait(&request, &status);

    // c)
    MPI_Sendrecv(&data[2-1+(DOMAINSIZE-1-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_bottom, 0, &data[2-1+(1-1)*DOMAINSIZE], SUBDOMAIN, MPI_DOUBLE, rank_top, 0, MPI_COMM_WORLD, &status);

    if (rank==9) {
        printf("data of rank 9 after communication\n");
        for (j=0; j<DOMAINSIZE; j++) {
            for (i=0; i<DOMAINSIZE; i++) {
                printf("%.1f ", data[i+j*DOMAINSIZE]);
            }
            printf("\n");
        }
    }

    MPI_Finalize();
    return 0;
}
